File: interview/lengthOfLIS.go
package interview

// ml(i) = ml(j) + 1
// j < i && nums[j] < nums[i] && nums[j]为最长的情况
func lengthOfLIS(nums []int) int {
	memo := make([]int, len(nums))
	memo[0] = 1
	for i := 1; i < len(nums); i++ {
		max := 0
		for j := 0; j < i; j++ {
			if nums[i] > nums[j] && memo[j] > max {
				max = memo[j]
			}
		}
		memo[i] = max + 1
	}
	return memo[len(nums)-1]
}

File: feishu/LengthOfLIS.java
/**
 * @author Zexho
 * @date 2022/3/26 9:00 PM
 */
public class LengthOfLIS {

    public int lengthOfLIS(int[] nums) {
        int[] memo = new int[nums.length + 1];
        memo[0] = Integer.MIN_VALUE;
        int idx = 0;
        for (int num : nums) {
            for (int j = idx; j >= 0; j--) {
                if (num > memo[j]) {
                    if (j == idx) {
                        memo[++idx] = num;
                    } else if (memo[j + 1] > num) {
                        memo[j + 1] = num;
                    }
                    break;
                }
            }
        }
        return idx;
    }

}
